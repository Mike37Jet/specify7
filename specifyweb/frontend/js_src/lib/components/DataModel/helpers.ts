import { f } from '../../utils/functools';
import type { RA } from '../../utils/types';
import { defined } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import type {
  AnySchema,
  AnyTree,
  SerializedModel,
  SerializedResource,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import { parseResourceUrl, resourceToJson } from './resource';
import { strictGetModel } from './schema';
import type { Tables } from './types';
import { addMissingFields } from './addMissingFields';

/** Like resource.toJSON(), but keys are converted to camel case */
export const serializeResource = <SCHEMA extends AnySchema>(
  resource: SerializedModel<SCHEMA> | SpecifyResource<SCHEMA>
): SerializedResource<SCHEMA> =>
  serializeModel<SCHEMA>(
    typeof resource.toJSON === 'function'
      ? resourceToJson(resource as SpecifyResource<SCHEMA>)
      : (resource as SerializedModel<SCHEMA>),
    (resource as SpecifyResource<SCHEMA>)?.specifyModel?.name
  );

export const specialFields = new Set([
  'id',
  'resource_uri',
  'recordset_info',
  '_tableName',
]);

/** Recursive helper for serializeResource */
function serializeModel<SCHEMA extends AnySchema>(
  resource: SerializedModel<SCHEMA>,
  tableName?: keyof Tables
): SerializedResource<SCHEMA> {
  const model = strictGetModel(
    defined(
      (tableName as SCHEMA['tableName']) ??
        resource._tableName ??
        parseResourceUrl((resource.resource_uri as string) ?? '')?.[0],
      `Unable to serialize resource because table name is unknown.` +
        (process.env.NODE_ENV === 'test'
          ? `\nMake sure your test file calls requireContext();`
          : '')
    )
  );
  const fields = [...model.fields.map(({ name }) => name), model.idField.name];

  return addMissingFields(
    model.name,
    Object.fromEntries(
      Object.entries(resource).map(([lowercaseFieldName, value]) => {
        let camelFieldName = fields.find(
          (fieldName) => fieldName.toLowerCase() === lowercaseFieldName
        );
        if (camelFieldName === undefined) {
          camelFieldName = lowercaseFieldName;
          if (!specialFields.has(lowercaseFieldName))
            console.warn(
              `Trying to serialize unknown field ${lowercaseFieldName} for table ${model.name}`,
              resource
            );
        }
        if (
          typeof value === 'object' &&
          value !== null &&
          !specialFields.has(camelFieldName)
        ) {
          const field = model.getField(lowercaseFieldName);
          const tableName =
            field === undefined || !field.isRelationship
              ? undefined
              : field.relatedModel.name;
          return [
            camelFieldName,
            Array.isArray(value)
              ? value.map((value) =>
                  serializeModel(
                    value as unknown as SerializedModel<SCHEMA>,
                    tableName
                  )
                )
              : serializeModel(value as SerializedModel<AnySchema>, tableName),
          ];
        } else return [camelFieldName, value];
      })
    )
  ) as SerializedResource<SCHEMA>;
}

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyModel.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyModel.name === tableName ? resource : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyModel.name) ? resource : undefined;
