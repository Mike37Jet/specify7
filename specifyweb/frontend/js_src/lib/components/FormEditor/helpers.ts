import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AttachmentOwnerBaseBusRules',
  'AttachmentReferenceBaseBusRules',
  'AuthorBusRules',
  'BaseTreeBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventAttributeBusRules',
  'CollectingEventAuthorizationBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectionSetupBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeaccessionBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DisposalAgentBusRules',
  'DisposalBusRules',
  'DisposalPreparationBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeInBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeoCoordDetailBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'InstitutionBusRules',
  'JournalBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'LocalityDetailBusRules',
  'PaleoContextBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'ReferenceWorkBusRules',
  'RepositoryAgreementBusRules',
  'ShipmentBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TableSearchResults',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeableSearchQueryBuilder',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
