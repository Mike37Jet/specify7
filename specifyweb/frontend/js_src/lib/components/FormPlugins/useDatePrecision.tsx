import React from 'react';

import { useResourceValue } from '../../hooks/useResourceValue';
import type { GetSet, RR } from '../../utils/types';
import type { AnySchema } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';

export const datePrecisions = { full: 1, 'month-year': 2, year: 3 } as const;
const reverseDatePrecisions: RR<number, PartialDatePrecision> = {
  1: 'full',
  2: 'month-year',
  3: 'year',
};
export type PartialDatePrecision = keyof typeof datePrecisions;

/** Fetch precision from the resource, and facilitate updating it */
export function useDatePrecision<SCHEMA extends AnySchema>(
  resource: SpecifyResource<SCHEMA> | undefined,
  precisionFieldName: (string & keyof SCHEMA['fields']) | undefined,
  defaultPrecision: PartialDatePrecision
): {
  readonly precision: GetSet<PartialDatePrecision>;
  readonly precisionValidationRef: React.RefCallback<HTMLSelectElement>;
} {
  const precisionField = React.useMemo(
    () =>
      precisionFieldName === undefined
        ? undefined
        : resource?.specifyTable.getField(precisionFieldName),
    [resource, precisionFieldName]
  );
  const numericDefaultPrecision = datePrecisions[defaultPrecision];
  const precisionParser = React.useMemo(
    () => ({ value: numericDefaultPrecision }),
    [numericDefaultPrecision]
  );
  const {
    value: numericPrecision,
    updateValue: setNumericPrecision,
    validationRef: precisionValidationRef,
  } = useResourceValue<keyof typeof reverseDatePrecisions>(
    resource,
    precisionField,
    precisionParser
  );
  const precision =
    (numericPrecision === undefined
      ? undefined
      : reverseDatePrecisions[numericPrecision]) ?? defaultPrecision;
  const setPrecision = React.useCallback(
    (precision: PartialDatePrecision) =>
      setNumericPrecision(datePrecisions[precision] ?? numericDefaultPrecision),
    [setNumericPrecision, numericDefaultPrecision]
  );

  return { precision: [precision, setPrecision], precisionValidationRef };
}

/**
 * Unset date precision if it is set, but date is not set.
 * Set date precision if it is not set, but date is set.
 *
 * @remarks
 * Precision may be out of sync in the database, or get out of sync when
 * date is set/cleared.
 */
export function useSyncDatePrecision(
  resource: SpecifyResource<AnySchema> | undefined,
  precisionFieldName: string | undefined,
  precision: PartialDatePrecision,
  hasDate: boolean | 'loading'
): void {
  const hasPrecision =
    typeof precisionFieldName === 'string' &&
    typeof resource?.get(precisionFieldName) === 'number';
  const hasDanglingPrecision = hasDate === false && hasPrecision;
  const needsPrecision =
    hasDate === true && typeof precisionFieldName === 'string' && !hasPrecision;
  const precisionToSet = hasDanglingPrecision
    ? null
    : needsPrecision
    ? datePrecisions[precision]
    : undefined;

  React.useEffect(
    () =>
      precisionToSet !== undefined && typeof precisionFieldName === 'string'
        ? void resource?.set(precisionFieldName, precisionToSet as never, {
            /*
             * Date precision may be in inconsistent state in the database. This
             * hook may fix it automatically on form load. That should not
             * trigger the save button
             */
            silent: true,
          })
        : undefined,
    [resource, precisionToSet, precisionFieldName]
  );
}
