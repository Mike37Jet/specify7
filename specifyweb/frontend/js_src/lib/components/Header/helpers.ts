export function formatDate(date: Date) {
  const year = date.getFullYear();
  const month = date.getMonth().toString();
  const day = date.getDate().toString();
  const hours = date.getHours().toString();
  const minutes = date.getMinutes().toString();
  const seconds = date.getSeconds().toString();

  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}
