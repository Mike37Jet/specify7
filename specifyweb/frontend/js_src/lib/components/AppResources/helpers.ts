import type { AppResourcesConformation } from './Aside';
import type { AppResources, AppResourcesTree } from './hooks';
import type { SpAppResource, SpViewSetObj } from '../DataModel/types';
import { f } from '../../utils/functools';
import type { SpecifyModel } from '../DataModel/specifyModel';
import type { RA } from '../../utils/types';
import { SerializedResource } from '../DataModel/helperTypes';

export const getAppResource = (
  resources: AppResources,
  model: SpecifyModel<SpAppResource | SpViewSetObj>,
  resourceId: number | undefined
): SerializedResource<SpAppResource | SpViewSetObj> | false | undefined =>
  typeof resourceId === 'number'
    ? (
        (model.name === 'SpAppResource'
          ? resources.appResources
          : resources.viewSets) as RA<SerializedResource<SpAppResource>>
      ).find(({ id }) => id === resourceId) ?? false
    : undefined;

export type AppResourceMode = 'appResources' | 'viewSets';

export const getAppResourceMode = (
  resource: SerializedResource<SpAppResource | SpViewSetObj>
): AppResourceMode =>
  resource._tableName === 'SpAppResource' ? 'appResources' : 'viewSets';

export const getAppResourceCount = (
  resourcesTree: AppResourcesTree[number]
): number =>
  resourcesTree.appResources.length +
  resourcesTree.viewSets.length +
  f.sum(resourcesTree.subCategories.map(getAppResourceCount));

export const buildAppResourceConformation = (
  resourcesTree: AppResourcesTree
): RA<AppResourcesConformation> =>
  resourcesTree
    .filter(
      ({ subCategories, appResources, viewSets }) =>
        subCategories.length > 0 ||
        appResources.length > 0 ||
        viewSets.length > 0
    )
    .map(({ key, subCategories }) => ({
      key,
      children: buildAppResourceConformation(subCategories),
    }));
