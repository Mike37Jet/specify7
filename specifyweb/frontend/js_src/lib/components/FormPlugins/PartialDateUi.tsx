import React from 'react';

import { useResourceValue } from '../../hooks/useResourceValue';
import { commonText } from '../../localization/common';
import { formsText } from '../../localization/forms';
import { dayjs, getDateInputValue } from '../../utils/dayJs';
import { f } from '../../utils/functools';
import { databaseDateFormat } from '../../utils/parser/dateConfig';
import { fullDateFormat, monthFormat } from '../../utils/parser/dateFormat';
import { parseDate } from '../../utils/parser/dayJsFixes';
import type { Parser } from '../../utils/parser/definitions';
import {
  getValidationAttributes,
  resolveParser,
} from '../../utils/parser/definitions';
import type { GetSet, RR } from '../../utils/types';
import { Button } from '../Atoms/Button';
import { Input, Select } from '../Atoms/Form';
import { dateParts } from '../Atoms/Internationalization';
import { ReadOnlyContext } from '../Core/Contexts';
import type { AnySchema } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { resourceOn } from '../DataModel/resource';
import { userPreferences } from '../Preferences/userPreferences';

export function isInputSupported(type: string): boolean {
  const input = document.createElement('input');
  const value = 'a';
  input.setAttribute('type', type);
  input.setAttribute('value', value);
  return input.value !== value;
}

const precisions = { full: 1, 'month-year': 2, year: 3 } as const;
const reversePrecisions: RR<number, PartialDatePrecision> = {
  1: 'full',
  2: 'month-year',
  3: 'year',
};
export type PartialDatePrecision = keyof typeof precisions;

/*
 * TESTS: this has been very buggy. add tests
 * REFACTOR: split this component into smaller
 */
export function PartialDateUi<SCHEMA extends AnySchema>({
  resource,
  dateFieldName,
  precisionField: precisionFieldName,
  defaultPrecision,
  defaultValue,
  id,
  canChangePrecision = true,
}: {
  readonly resource: SpecifyResource<SCHEMA> | undefined;
  readonly dateFieldName: string & keyof SCHEMA['fields'];
  readonly precisionField: (string & keyof SCHEMA['fields']) | undefined;
  readonly defaultPrecision: PartialDatePrecision;
  readonly defaultValue: Date | undefined;
  readonly id: string | undefined;
  readonly canChangePrecision?: boolean;
}): JSX.Element {
  // Preferences
  const [useDatePicker] = userPreferences.use(
    'form',
    'ui',
    'useAccessibleFullDatePicker'
  );
  const [useMonthPicker] = userPreferences.use(
    'form',
    'ui',
    'useAccessibleMonthPicker'
  );
  const {
    dateType,
    dateSupported,
    monthType,
    monthSupported,
    inputFullFormat,
    inputMonthFormat,
  } = React.useMemo(() => {
    const dateType = useDatePicker ? 'date' : 'text';
    const monthType = useMonthPicker ? 'month' : 'text';
    const dateSupported = useDatePicker && isInputSupported('date');
    const monthSupported = useMonthPicker && isInputSupported('month');

    return {
      dateType,
      dateSupported,
      monthType,
      monthSupported,
      /*
       * If input[type="date"] or input[type="month"] is not supported,
       * present the date in a more human-readable format
       */
      inputFullFormat: dateSupported ? databaseDateFormat : fullDateFormat(),
      inputMonthFormat: monthSupported ? 'YYYY-MM' : monthFormat(),
    };
  }, [useDatePicker, useMonthPicker]);

  const {
    precision: [precision, setPrecision],
    precisionValidationRef,
  } = useDatePrecision(resource, precisionFieldName, defaultPrecision);

  // Value field
  const dateField = React.useMemo(
    () =>
      dateFieldName === undefined
        ? undefined
        : resource?.specifyTable.getField(dateFieldName),
    [resource, dateFieldName]
  );
  const parser = React.useMemo<Parser>(
    () => ({
      value: f.maybe(defaultValue, getDateInputValue),
    }),
    [defaultValue, precision]
  );
  const validationAttributes = React.useMemo(
    () =>
      precision === 'month-year'
        ? {}
        : getValidationAttributes({
            ...resolveParser(dateField ?? {}, {
              type: precision === 'full' ? 'java.util.Date' : precision,
            }),
            ...parser,
          }),
    [dateField, parser, precision]
  );
  const {
    value: inputValue = '',
    updateValue: setInputValue,
    inputRef,
    validationRef,
    setValidation,
  } = useResourceValue<string | null>(resource, dateField, parser);

  const syncMoment = React.useCallback(
    (moment: ReturnType<typeof dayjs> | undefined) => {
      const value = resource?.get(dateFieldName) ?? undefined;
      const newMoment =
        value === undefined
          ? undefined
          : dayjs(value, databaseDateFormat, true);

      return moment === undefined ||
        newMoment === undefined ||
        moment.toJSON() !== newMoment.toJSON()
        ? newMoment
        : moment;
    },
    [resource, dateFieldName]
  );

  // Parsed date object
  const [moment, setMoment] = React.useState<
    ReturnType<typeof dayjs> | undefined
    /*
     * Can't set initialState here because it won't be reEvaluated when
     * the "resource" changes
     */
  >(undefined);

  const isInitialized = React.useRef<boolean>(false);

  React.useEffect(() => {
    if (resource === undefined) return;
    isInitialized.current = false;

    return resourceOn(
      resource,
      `change:${dateFieldName}`,
      () => setMoment(syncMoment),
      true
    );
  }, [
    resource,
    dateFieldName,
    precisionFieldName,
    defaultPrecision,
    defaultValue,
    syncMoment,
  ]);

  const isReadOnly = React.useContext(ReadOnlyContext);
  React.useEffect(() => {
    if (resource === undefined) return;
    /*
     * If resource changes, a new moment is set, but its value won't get
     * propagated on the first call to this useEffect.
     * It is demonstrated here: https://codepen.io/maxpatiiuk/pen/oNqNqVN
     */
    if (!isInitialized.current) {
      isInitialized.current = true;
      return;
    }
    if (moment === undefined) {
      resource.set(dateFieldName, null as never);
      if (
        precisionFieldName !== undefined &&
        typeof resource.get(precisionFieldName) !== 'number'
      )
        resource.set(precisionFieldName, null as never, {
          silent: true,
        });
      setValidation(formsText.invalidDate());
      setInputValue('');
    } else if (moment.isValid()) {
      const value = moment.format(databaseDateFormat);

      if (
        precisionFieldName !== undefined &&
        typeof resource.get(precisionFieldName) !== 'number'
      )
        resource.set(precisionFieldName, precisions[precision] as never, {
          silent: true,
        });

      if (!isReadOnly) {
        const oldRawDate = resource.get(dateFieldName);
        const oldDate =
          typeof oldRawDate === 'string' ? new Date(oldRawDate) : undefined;
        const newDate = moment.toDate();
        /*
         * Back-end may return a date in a date-time format, which when converted
         * to date format causes front-end to trigger a needless unload protect
         * See https://github.com/specify/specify7/issues/2578. This fixes that
         */
        const isChanged =
          f.maybe(oldDate, getDateInputValue) !== getDateInputValue(newDate);
        resource.set(dateFieldName, value as never, { silent: !isChanged });
      }
      setValidation('');

      if (precision === 'full') setInputValue(moment.format(inputFullFormat));
      else if (precision === 'month-year')
        setInputValue(moment.format(inputMonthFormat));
      else setInputValue(moment.year().toString());
    } else {
      const validationMessage =
        precision === 'full'
          ? formsText.requiredFormat({ format: fullDateFormat() })
          : precision === 'month-year'
          ? formsText.requiredFormat({ format: monthFormat() })
          : formsText.invalidDate();
      setValidation(validationMessage);
    }
  }, [
    isReadOnly,
    setValidation,
    setInputValue,
    resource,
    moment,
    precision,
    dateFieldName,
    precisionFieldName,
    inputFullFormat,
    inputMonthFormat,
  ]);

  function handleChange(initialValue?: string): void {
    const input = inputRef.current;
    if (isReadOnly || input === null) return;

    const value = initialValue ?? input.value.trim();

    setMoment(value.length > 0 ? parseDate(precision, value) : undefined);
  }

  return (
    <div className="flex w-full gap-1">
      {!isReadOnly && canChangePrecision ? (
        <label>
          <span className="sr-only">{formsText.datePrecision()}</span>
          <Select
            className="!w-auto !min-w-[unset] print:hidden"
            forwardRef={precisionValidationRef}
            value={precision}
            onBlur={(): void => {
              if (moment === undefined) return;
              let newMoment = dayjs(moment);
              if (precision === 'year' || precision === 'month-year')
                newMoment = newMoment.date(1);
              if (precision === 'year') newMoment = newMoment.month(0);

              /*
               * This avoids the following message in the console:
               * "The specified value does not conform to the required format"
               */
              setInputValue('');

              setMoment(newMoment);
            }}
            onChange={({ target }): void => {
              if (resource === undefined) return;
              const precision = target.value as PartialDatePrecision;
              setPrecision(precision);
              const precisionIndex = precisions[precision];
              if (
                typeof moment === 'object' &&
                typeof precisionFieldName === 'string'
              )
                resource.set(precisionFieldName, precisionIndex as never);
            }}
          >
            <option value="full">{commonText.fullDate()}</option>
            <option value="month-year">{formsText.monthYear()}</option>
            <option value="year">{dateParts.year}</option>
          </Select>
        </label>
      ) : undefined}
      <Input.Generic
        forwardRef={validationRef}
        id={id}
        isReadOnly={isReadOnly}
        value={inputValue ?? ''}
        onBlur={f.zero(handleChange)}
        onValueChange={setInputValue}
        {...(precision === 'year'
          ? {
              ...validationAttributes,
              placeholder: formsText.yearPlaceholder(),
            }
          : {
              ...(precision === 'month-year'
                ? {
                    type: monthType,
                    placeholder: monthFormat(),
                    title: moment?.format(monthFormat()),
                    ...(monthSupported
                      ? {}
                      : {
                          minLength: monthFormat().length,
                          maxLength: monthFormat().length,
                        }),
                  }
                : {
                    type: dateType,
                    placeholder: fullDateFormat(),
                    title: moment?.format(fullDateFormat()),
                    min: validationAttributes.min,
                    max: validationAttributes.max,
                  }),
            })}
      />
      {!isReadOnly &&
      ((precision === 'full' && !dateSupported) ||
        (precision === 'month-year' && !monthSupported)) ? (
        <Button.Icon
          aria-label={formsText.today()}
          icon="calendar"
          title={formsText.todayButtonDescription()}
          onClick={(): void => setMoment(dayjs())}
        />
      ) : undefined}
    </div>
  );
}

function useDatePrecision<SCHEMA extends AnySchema>(
  resource: SpecifyResource<SCHEMA> | undefined,
  precisionFieldName: (string & keyof SCHEMA['fields']) | undefined,
  defaultPrecision: PartialDatePrecision
): {
  readonly precision: GetSet<PartialDatePrecision>;
  readonly precisionValidationRef: React.RefCallback<HTMLSelectElement>;
} {
  const precisionField = React.useMemo(
    () =>
      precisionFieldName === undefined
        ? undefined
        : resource?.specifyTable.getField(precisionFieldName),
    [resource, precisionFieldName]
  );
  const precisionParser = React.useMemo(
    () => ({ value: defaultPrecision }),
    [defaultPrecision]
  );
  const {
    value: numericPrecision,
    updateValue: setNumericPrecision,
    validationRef: precisionValidationRef,
  } = useResourceValue<keyof typeof reversePrecisions>(
    resource,
    precisionField,
    precisionParser
  );
  const precision =
    (numericPrecision === undefined
      ? undefined
      : reversePrecisions[numericPrecision]) ?? defaultPrecision;
  const setPrecision = React.useCallback(
    (precision: PartialDatePrecision) =>
      setNumericPrecision(precisions[precision] ?? defaultPrecision),
    [setNumericPrecision, defaultPrecision]
  );
  return { precision: [precision, setPrecision], precisionValidationRef };
}
