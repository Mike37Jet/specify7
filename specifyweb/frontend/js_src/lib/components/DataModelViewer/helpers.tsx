import React from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { schemaText } from '../../localization/schema';
import { syncFieldFormat } from '../../utils/fieldFormat';
import { f } from '../../utils/functools';
import { resolveParser } from '../../utils/parser/definitions';
import { ensure, IR, RA, RR } from '../../utils/types';
import { formatNumber } from '../Atoms/Internationalization';
import { getField } from '../DataModel/helpers';
import { schema } from '../DataModel/schema';
import { TableIcon } from '../Molecules/TableIcon';

export type Value =
  | number
  | string
  | readonly [number | string | undefined, JSX.Element]
  | undefined;
export type Row<SHAPE extends IR<Value>> = SHAPE;

const parser = f.store(() =>
  resolveParser(
    {},
    {
      type: 'java.lang.Boolean',
    }
  )
);

export const booleanFormatter = (value: boolean): string =>
  syncFieldFormat(undefined, parser(), value);

export const tableColumns = f.store(
  () =>
    ({
      name: getField(schema.models.SpLocaleContainer, 'name').label,
      label: schemaText.fieldLabel(),
      isSystem: getField(schema.models.SpLocaleContainer, 'isSystem').label,
      isHidden: getField(schema.models.SpLocaleContainer, 'isHidden').label,
      tableId: schemaText.tableId(),
      fieldCount: schemaText.fieldCount(),
      relationshipCount: schemaText.relationshipCount(),
    } as const)
);

export const getTables = () =>
  ensure<RA<Row<RR<keyof ReturnType<typeof tableColumns>, Value>>>>()(
    Object.values(schema.models).map(
      (model) =>
        ({
          name: [
            model.name.toLowerCase(),
            <>
              <TableIcon label={false} name={model.name} />
              {model.name}
            </>,
          ],
          label: model.label,
          isSystem: booleanFormatter(model.isSystem),
          isHidden: booleanFormatter(model.isHidden),
          tableId: [
            model.tableId,
            <span className="flex w-full justify-end tabular-nums">
              {model.tableId}
            </span>,
          ],
          fieldCount: [
            model.fields.length,
            <span className="flex w-full justify-end tabular-nums">
              {formatNumber(model.fields.length)}
            </span>,
          ],
          relationshipCount: [
            model.relationships.length,
            <span className="flex w-full justify-end tabular-nums">
              {formatNumber(model.relationships.length)}
            </span>,
          ],
        } as const)
    )
  );

export function DataModelRedirect(): null {
  const { tableName = '' } = useParams();
  const navigate = useNavigate();
  React.useEffect(() => {
    navigate(`/specify/data-model/#${tableName}`, { replace: true });
  }, []);
  return null;
}
