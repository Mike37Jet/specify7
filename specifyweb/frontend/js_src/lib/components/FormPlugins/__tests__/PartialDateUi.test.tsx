import { mockTime, requireContext } from '../../../tests/helpers';
import { snapshot } from '../../../tests/reactUtils';
import type { SpecifyResource } from '../../DataModel/legacyTypes';
import { tables } from '../../DataModel/tables';
import type { CollectionObject } from '../../DataModel/types';
import { PartialDateUi } from '../PartialDateUi';
import { datePrecisions } from '../useDatePrecision';

mockTime();
requireContext();

// Test helpers
const dateFieldName = 'catalogedDate';
const precisionField = 'catalogedDatePrecision';
const getBaseResource = (): SpecifyResource<CollectionObject> =>
  new tables.CollectionObject.Resource(undefined, { noBusinessRules: true });
function getResource(): SpecifyResource<CollectionObject> {
  const resource = getBaseResource();
  resource.set(dateFieldName, '2023-12-07');
  resource.set(precisionField, datePrecisions.full);
  return resource;
}

const baseProps: Parameters<typeof PartialDateUi>[0] = {
  resource: undefined,
  dateFieldName,
  precisionField,
  defaultPrecision: 'full',
  defaultValue: undefined,
  id: undefined,
  isRequired: false,
  canChangePrecision: true,
};
const props = (
  resource = getResource(),
  extra: Partial<Parameters<typeof PartialDateUi>[0]> = {}
): Parameters<typeof PartialDateUi>[0] => ({
  ...baseProps,
  resource,
  ...extra,
});

// Snapshot tests
snapshot(PartialDateUi, () => props(getBaseResource()));
snapshot(PartialDateUi, () =>
  props(getBaseResource(), {
    canChangePrecision: false,
    defaultPrecision: 'month-year',
  })
);
snapshot(PartialDateUi, () =>
  props(getBaseResource(), {
    canChangePrecision: false,
    defaultPrecision: 'year',
  })
);

// Interaction tests

/*
 * Test all props with all precision types?
 * Test reading existing values from resource
 * Test non-accessible date pickers
 * Test changing precision on the fly and how that mutates value
 * Refactor the date picker into smaller, more testable parts?
 */

/*
 *Test('<Form> removes notSubmitted class name on submit', async () => {
 *const { asFragment, getByRole, user } = mount(
 *  <Form aria-label="form" onSubmit={(event): void => event.preventDefault()}>
 *    <input type="submit" />
 *  </Form>
 *);
 *const form = getByRole('form');
 *expect(form).toHaveClass(className.notSubmittedForm);
 *
 *const button = getByRole('button');
 *await user.click(button);
 *
 *expect(form).not.toHaveClass(className.notSubmittedForm);
 *
 *expect(asFragment()).toMatchSnapshot();
 *});
 */
