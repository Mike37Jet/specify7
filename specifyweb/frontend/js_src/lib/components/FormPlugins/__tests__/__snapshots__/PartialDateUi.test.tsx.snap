// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PartialDateUi renders without errors 1`] = `
<DocumentFragment>
  <div
    class="flex w-full gap-1"
  >
    <label>
      <span
        class="sr-only"
      >
        Date Precision
      </span>
      <select
        class="not-touched w-full pr-5 bg-right cursor-pointer !w-auto !min-w-[unset] print:hidden"
      >
        <option
          value="full"
        >
          Full Date
        </option>
        <option
          value="month-year"
        >
          Mon / Year
        </option>
        <option
          value="year"
        >
          Year
        </option>
      </select>
    </label>
    <input
      class="not-touched w-full"
      max="9999-12-31"
      placeholder="MM/DD/YYYY"
      title="08/31/2022"
      type="date"
      value="2022-08-31"
    />
  </div>
</DocumentFragment>
`;

exports[`PartialDateUi renders without errors 2`] = `
<DocumentFragment>
  <div
    class="flex w-full gap-1"
  >
    <input
      class="not-touched w-full"
      placeholder="MM/YYYY"
      title="08/2022"
      type="month"
      value="2022-08-01"
    />
  </div>
</DocumentFragment>
`;

exports[`PartialDateUi renders without errors 3`] = `
<DocumentFragment>
  <div
    class="flex w-full gap-1"
  >
    <input
      class="not-touched w-full"
      max="9999"
      min="1"
      placeholder="YYYY"
      step="1"
      type="number"
      value="2022-01-01"
    />
  </div>
</DocumentFragment>
`;
